rts/Rml/Backends/RmlUi_Backend.cpp

```diff
@@ -0,0 +1,495 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+#include <RmlUi/Core.h>
+#include <RmlUi/Core/Profiling.h>
+#include <RmlUi/Debugger.h>
+#include <SDL.h>
+#include <functional>
+#include <tracy/Tracy.hpp>
+
+#include "Lua/LuaUI.h"
+#include "Rendering/Textures/Bitmap.h"
+#include "Rml/Components/ElementLuaTexture.h"
+#include "Rml/RmlInputReceiver.h"
+#include "Rml/SolLua/RmlSolLua.h"
+#include "RmlUi_Backend.h"
+
+#ifndef HEADLESS
+#include "RmlUi_Renderer_GL3_Recoil.h"
+#else
+#include "RmlUi_Renderer_Headless.h"
+#endif
+
+#include "RmlUi_SystemInterface.h"
+#include "RmlUi_VFSFileInterface.h"
+#include "System/Input/InputHandler.h"
+#include "System/Log/ILog.h"
+
+using CtxMutex = std::recursive_mutex;
+using CtxLockGuard = std::lock_guard<CtxMutex>;
+struct lua_State;
+
+/// Passes through RML events to the function pointers given in the constructor
+class PassThroughPlugin : public Rml::Plugin
+{
+	void (*onContextCreate)(Rml::Context*);
+	void (*onContextDestroy)(Rml::Context*);
+
+public:
+	PassThroughPlugin(void (*onContextCreate)(Rml::Context*),
+	                  void (*onContextDestroy)(Rml::Context*))
+		: onContextCreate{onContextCreate}
+		, onContextDestroy{onContextDestroy}
+	{
+	}
+
+	int GetEventClasses() override
+	{
+		return EVT_BASIC;
+	}
+
+	void OnInitialise() override{};
+	void OnShutdown() override{};
+
+	void OnContextCreate(Rml::Context* context) override
+	{
+		onContextCreate(context);
+	};
+
+	void OnContextDestroy(Rml::Context* context) override
+	{
+		onContextDestroy(context);
+	};
+};
+
+/**
+  Global data used by this backend.
+  Lifetime governed by the calls to Backend::Initialize() and
+  Backend::Shutdown().
+ */
+struct BackendData {
+	RmlSystemInterface system_interface;
+#ifndef HEADLESS
+	RenderInterface_GL3_Recoil render_interface;
+#else
+	RenderInterface_Headless render_interface;
+#endif
+	VFSFileInterface file_interface;
+
+	SDL_Window* window = nullptr;
+	SDL_GLContext gl_context = nullptr;
+	std::vector<Rml::Context*> contexts;
+	InputHandler::SignalType::connection_type inputCon;
+	CRmlInputReceiver inputReceiver;
+
+	// make atomic_bool?
+	bool initialized = false;
+	bool debuggerAttached = false;
+	int winX = 1;
+	int winY = 1;
+
+	lua_State* ls = nullptr;
+	Rml::SolLua::SolLuaPlugin* luaPlugin = nullptr;
+
+	CtxMutex contextMutex;
```

Comment:
Generally regarding `CtxMutex` and concurrency mechanisms and thread safety:

@ChrisFloofyKitsune Are you an expert on concurrency and thread safety, and are you intimately familiar with how threading is done in BAR? If the answers to those questions are not both 'yes', I would very strongly recommend that you find anyone in BAR who understands the threading model of BAR well and help them guide you and review all concurrency-related parts. Concurrency and thread is something to work with without understanding what you are doing. I assume that there is a developer in BAR that understands BAR's concurrency model. Or if this part of BAR is a fork of Spring, then ask for help from a relevant developer in Spring that understands the engine's concurrency model. Concurrency is a nebulous topic, and many developers get it completely wrong or do not even understand the basics, even "senior engineers".

If you can guarantee that RmlUi is only ever accessed and worked with (and the same for any resources accessed by RmlUi) by the same specific thread over the course of the process' lifetime, then my belief is that there should not be any issues. In such a case, it may even be possible to omit the `CtxMutex`, since if it is only the same specific thread that ever accesses and works with RmlUi and any resources accessed by RmlUi, RmlUi would effectively be single-threaded as my understanding goes.

Reply:
@sprunk came with in my opinion a good argument for reasoning about this topic in https://discord.com/channels/549281623154229250/724924957074915358/1226393299338924062 . I do not understand or know enough about LuaUI to determine whether LuaUI and the current RmlUi integration touches/interacts with only the same things. Does anyone else know LuaUI sufficiently well to determine whether that's the case for this PR? If it is, I would believe that the mutexes can be removed without issue. Keeping things single-threaded (when possible) is one of the easier and safer ways of dealing with threads.

Reply:
before adding the mutexes, I was able to trigger a crash by rapidly toggling on and off a widget that created a RmlUi context in Initialize() and tore it down in Shutdown()

Reply:
Unless I am much mistaken, that indicates that RmlUi is accessed from multiple threads. According to https://mikke89.github.io/RmlUiDoc/pages/faq.html :

> Is RmlUi thread safe?
>
> No guarantees are provided in terms of thread safety. Any access into the RmlUi API from multiple threads need to be synchronized such as by using a mutex.

As far as I can figure out, this possibly means that any access to RmlUi should be protected with a mutex, not only when accessing a context or manipulating a list of contexts.

Your context guard does help from what I can tell, but it might well not be sufficient to make it overall thread safe.

I wonder why multiple threads are involved. But I do not know or understand the BAR/Spring engine. Is it possible to avoid having multiple threads involved? I will ask on Discord, hopefully someone can help give pointers.

Reply:
My suspicion is the UI rendering and Lua updating may be on separate threads?

Reply:
I wonder if basic logging that includes something like https://en.cppreference.com/w/cpp/thread/thread/get_id would help confirm whether multiple threads are involved.

I am tempted to give this a go, but I have never built BAR. p2004a have given instructions somewhere on building this PR, which I guess that I would be able to figure out (and I can always ask in Discord), but which specific steps would I have to take to reproduce it?

It would be nice if you could try to build, and remove some of the locks, and see if you can still reproduce it. But, getting good and enough sleep is important, one should remember to take good care of oneself, it is important.

Reply:
This have been resolved :+1: .